mlir/lib/Dialect/Rock/Transforms/RemoveRedundantCasts.cpp

```diff
@@ -0,0 +1,510 @@
+             "Output element type must be smaller than f32");
+    } else {
+      llvm_unreachable("Unsupported element type for truncation");
+    }
+
+    // Create indexing maps - both input and output use the same identity map
+    MLIRContext *ctx = rewriter.getContext();
+    int64_t rank = inputType.getRank();
+    SmallVector<AffineExpr> exprs;
+    for (int64_t i = 0; i < rank; ++i) {
+      exprs.push_back(getAffineDimExpr(i, ctx));
+    }
+    auto identityMap = AffineMap::get(rank, 0, exprs, ctx);
+    SmallVector<AffineMap> indexingMaps = {identityMap, identityMap};
+
+    // Create iterator types - all parallel
+    SmallVector<utils::IteratorType> iteratorTypes(
+        rank, utils::IteratorType::parallel);
+
+    // Create the linalg.generic operation
+    auto loc = input->getLoc();
+    auto genericOp = rewriter.create<linalg::GenericOp>(
+        loc,
+        /*resultTypes=*/TypeRange{outputType},
+        /*inputs=*/ValueRange{input->getResult(0)},
+        /*outputs=*/ValueRange{output->getResult(0)},
+        /*indexingMaps=*/indexingMaps,
+        /*iteratorTypes=*/iteratorTypes,
+        /*doc=*/"",
+        /*library_call=*/"",
+        /*bodyBuild=*/
+        [](OpBuilder &builder, Location loc, ValueRange args) {
+          Value blockArg = args[0];
+          Value outputArg = args[1];
+          Type oType = outputArg.getType();
+          Value truncResult =
+              builder.create<arith::TruncFOp>(loc, oType, blockArg);
+          builder.create<linalg::YieldOp>(loc, truncResult);
+        });
+
+    return genericOp;
+  }
+
+  bool handleRockGemmWrapper(Operation *rockOp, linalg::GenericOp extGen,
+                             PatternRewriter &rewriter) const {
+    LLVM_DEBUG(llvm::dbgs()
+               << "\tFound "
+               << "RockGemmWrapperInterface/RockGemmGemmWrapperInterface: "
+               << *rockOp << "\n");
+
+    Value extGenResult = extGen.getResult(0);
+    Value extGenInput = extGen.getInputs()[0];
+
+    auto extOutputType = dyn_cast<RankedTensorType>(extGenResult.getType());
+    auto extInputType = dyn_cast<RankedTensorType>(extGenInput.getType());
+
+    if (!extOutputType || !extInputType) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tOne or more types are not RankedTensorType, skipping\n");
+      return false;
+    }
+
+    Type rockOutputElementType;
+    Value rockOutputOp;
+    if (auto rockI = dyn_cast<RockGemmWrapperInterface>(rockOp)) {
+      rockOutputElementType = rockI.getCType();
+      rockOutputOp = rockI.getOutArgument()->get();
+    } else if (auto rockGemmI =
+                   dyn_cast<RockGemmGemmWrapperInterface>(rockOp)) {
+      rockOutputElementType = rockGemmI.getCType();
+      rockOutputOp = rockGemmI.getOutArgument()->get();
+    } else {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tNot a RockGemmWrapperInterface, skipping\n");
+      return false;
```

Comment:
This should be returning `llvm_fatal_error`. This function is only meant to be called on Gemm Based ops. Or else you can do this check at the very beginning of the function.